rfd/0182-multi-port-tcp-app-access.md

```diff
@@ -0,0 +1,544 @@
+an error in Connect.
+
+###### Invalid app spec
+
+Alice wants to extend the app spec with another port, 6060, but she makes a mistake and forgets a
+comma:
+
+```yaml
+apps:
+- name: "waldo"
+  uri: "tcp://localhost"
+  ports: [4080-4090, 50956060]
+  labels:
+    env: prod
+```
+
+The relevant UI (`teleport start`, `tctl edit`, etc.) returns an error about an incorrect port
+in the app spec and prevents Alice from saving such app spec.
+
+### Passing the port number from the client to the app agent
+
+In order to pass the port number from the client to the app service, the port is going to be
+included in `RouteToApp` portion of the app cert. `RouteToApp` is used today to route connections to
+single-port TCP apps. With a few modifications, it can be used for multi-port access.
+
+`RouteToApp` already has the `URI` field which includes the port number. However, the `URI` field is
+a recent addition used solely for audit events, not as the source of truth in routing. Instead, the
+app service reads [the current URI from the cache of the backend state](https://github.com/gravitational/teleport/blob/d6640d89d8a99908d7ac5385f2cfe0e08e230239/lib/srv/app/tcpserver.go#L45-L55).
+A valid app cert grants access to a specific app as it's defined at any given moment on the backend,
+not as it was defined at the time of generating a cert. We want to keep this behavior for
+single-port apps after introducing multi-port apps.
+
+As such, we cannot reuse the `URI` field by simply changing the port number in it. When looking at a
+cert like this during routing, it'd be impossible to tell these two scenarios apart:
+
+* The cert was issued for a multi-port app and allows access to a single port.
+* The cert was issued for a single-port app and allows access to the port currently defined in the
+  app spec.
+
+The `URI` field should be left unchanged to merely mirror the same field from the app spec.
+
+Instead, we add a `TargetPort` field to `RouteToApp`. The presence of this field signifies that the
+cert grants access to the specific port of a specific app, as long as the port is currently defined
+in the app spec. The `URI` field still won't be used as the source of truth for the hostname during
+routing as we continue to use `URI` from the app spec.
+
+When the port makes its way to the app service, the app service checks the port against ports
+defined in the app spec before deciding whether to proxy the connection or not. The app service
+[already has access to the app spec](https://github.com/gravitational/teleport/blob/5012036ad28a25d54858e454364828803f558e0b/lib/srv/app/connections_handler.go#L726).
+
+At this point we don't plan to introduce RBAC for port numbers, but this is something we can
+consider in the future. Port numbers on an individual app are akin to database names within a single
+database server. Customers might want to treat them as such, by e.g. limiting access to different
+ports.
```

Comment:
aside: i think there's a workaround where you register the app twice with different ports and allow different roles to access each app
```yaml
- name: foo-dev
  public_addr: foo.example.com
  labels:
    team: devs
  uri: tcp://localhost
  ports:
  - start: 443
  - start: 8080
- name: foo-admin
  public_addr: foo.example.com
  labels:
    team: admins
  uri: tcp://localhost
  ports:
  - start: 1
    end: 65535
```

Reply:
I'm not sure how well this would work, given that the app service looks for an app using its `public_addr` and not `name`.

<details>
<summary>Relevant code</summary>

https://github.com/gravitational/teleport/blob/5a7e73c91a4ef75998df2079f38d262c4f480374/lib/srv/app/connections_handler.go#L726

https://github.com/gravitational/teleport/blob/5a7e73c91a4ef75998df2079f38d262c4f480374/lib/srv/app/server.go#L549-L564

https://github.com/gravitational/teleport/blob/5a7e73c91a4ef75998df2079f38d262c4f480374/lib/service/service.go#L4547-L4559
</details>

Reply:
ah, never mind then